src/grade_recognition_systems.asm

```diff
@@ -0,0 +1,383 @@
+    db 8 
+    db 8 
+    db 9 
+    db 9 
+    db 10 
+    db 11 
+    db 12 
+    db 12 
+    db 12 
+    db 13 
+    db 13 
+    db 14 
+    db 14 
+    db 15
+    db 15 
+    db 16 
+    db 16 
+    db 17
+sComboMultipliers:
+    db 1, 1, 1, 1, 1 ; Combo size, (Multiplier for: ) Single, Double, Triple, Tetris
+    db 2, 1, 1.2, 1.4, 1.5
+    db 3, 1, 1.2, 1.5, 1.8
+    db 4, 1, 1.4, 1.6, 2
+    db 5, 1, 1.4, 1.7, 2.2
+    db 6, 1, 1.4, 1.8, 2.3
+    db 7, 1, 1.4, 1.9, 2.4
+    db 8, 1, 1.5, 2, 2.5
+    db 9, 1, 1.5, 2.1, 2.6
+    db 10, 2, 2.5, 3
+sLevelMultiplier:
+    db 1 ; 000-249
+    db 2 ; 250-499
+    db 3 ; 500-749
+    db 4 ; 750-999
+sBaselineCOOL:
+    db 52 ;070 (value in seconds)
+    db 52 ;170
+    db 49 ;270
+    db 45 ;370
+    db 45 ;470
+    db 42 ;570
+    db 42 ;670
+    db 38 ;770
+    db 38 ;870
+sREGRETConditions:
+    db 1, 30 ;minutes, seconds
+    db 1, 15
+    db 1, 15 
+    db 1, 8 
+    db 1, 0
+    db 1, 0 
+    db 50 ;seconds
+    db 50 
+    db 50 
+    db 50
+
+SECTION "Grading Functions", ROM0
+    ; Wipe the grading variables.
+GradeInit::
+    xor a, a
+    ld [wDecayRate], a
+    ld [wGradePoints], a
+    ld [wInternalGrade], a
+    ld [wDisplayedGrade], a
+    ld a, $7
+    ld [wPalette], a
+    jp UpdateGrade
+    ret
+
+
+    ; Gets the highest grade the player qualifies for.
+UpdateGrade::
+    ; Is the Speed Curve Death or Shirase?
+    ld hl, .GradeJumps
+    ld a, [wSpeedCurveState]
+    sub a, $3
+    cp a, $3
+    jp nc, .Normal
+    ld b, 0
+    cp a, $1
+    jr c, :+
+    inc a
+    inc a
+:   cp a, $4
+    jr nc, .Normal
+    ld c, a
+    add hl, bc
+    jp hl
+.GradeJumps
+    jp Death
+    jp Shirase
+.Normal
+    ; Skip to GM check if past S9.
+    ld a, [wDisplayedGrade]
+    cp a, GRADE_S9
+    jr c, .trygradeup
+    jr CheckForGM
+
+.trygradeup
+    ; Get our score into BC
+    call PrepareScore
+
+    ; Double our current grade and use it as an offset into the scoring table.
+    ld a, [wDisplayedGrade]
+    add a
+    ld d, 0
+    ld e, a
+
+    ; Have HL point to the next required score and get it into DE.
+    ld hl, sGradeScores
+    add hl, de
+
+    ; LSB
+    ld a, [hl+]
+    ld e, a
+
+    ; MSB
+    ld a, [hl]
+    ld d, a
+
+    ; Check if BC >= DE...
+    ; Return if B < D.
+    ld a, b
+    cp a, d
+    ret c
+
+    ; We can confidently increase the grade if B > D.
+    jr nz, .increasegrade
+
+    ; If B == D, we need to check C and E...
+
+    ; Return if C < E. Otherwise increase the grade.
+    ld a, c
+    cp a, e
+    ret c
+
+.increasegrade
+    ; Add 1 to the grade.
+    ld a, [wDisplayedGrade]
+    inc a
+    ld [wDisplayedGrade], a
+    ; Prepare the effect stuff
+    ld a, $f
+    ld [wEffectTimer], a
+    xor a, a
+    ld [wPalette], a
+
+    ; Play the jingle.
+    ld a, SFX_RANKUP
+    call SFXEnqueue
+
+    ; Loop and see if we can increment more grades.
+    ld a, [wDisplayedGrade]
+    cp a, GRADE_S9
+    ret z
+    jr .trygradeup
+
+
+CheckForGM:

```

Comment:
Should be renamed to something like CheckForTGM1GM since this one is only used for UpdateGradeTGM1